File: server/src/main/java/com/antigravity/converters/TrackmateConfigConverter.java
package com.antigravity.converters;

import com.antigravity.protocols.trackmate.TrackmateConfig;

public class TrackmateConfigConverter {

  public static com.antigravity.proto.TrackmateConfig toProto( // fqn-collision
      TrackmateConfig config) { // fqn-collision
    if (config == null) {
      return com.antigravity.proto.TrackmateConfig.getDefaultInstance(); // fqn-collision
    }
    return com.antigravity.proto.TrackmateConfig.newBuilder() // fqn-collision
        .setName(config.name != null ? config.name : "")
        .setCommPort(config.commPort != null ? config.commPort : "")
        .setNormallyClosedRelays(config.normallyClosedRelays)
        .setUseIr(config.useIR)
        .setDebounce(config.debounce)
        .setNumLanes(config.numLanes)
        .build();
  }
}

File: client/src/app/converters/trackmate_config.converter.ts
import { TrackmateConfig } from "@app/models/track";
import { ITrackmateConfig } from "@app/proto/antigravity";

export class TrackmateConfigConverter {
  static fromProto(proto: ITrackmateConfig): TrackmateConfig {
    return {
      name: proto.name || "Trackmate",
      commPort: proto.commPort || "",
      normallyClosedRelays: proto.normallyClosedRelays ?? true,
      normallyClosedLaneSensors: proto.normallyClosedLaneSensors ?? true,
      useIR: proto.useIr ?? true,
      debounce: proto.debounce || 1,
      numLanes: proto.numLanes || 8,
      hasPerLaneRelays: proto.hasPerLaneRelays ?? false,
      lapPinPitBehavior: proto.lapPinPitBehavior || 0,
      lapPinBehaviors: proto.lapPinBehaviors || [],
    };
  }

  static toProto(config: TrackmateConfig): ITrackmateConfig {
    return {
      name: config.name,
      commPort: config.commPort,
      normallyClosedRelays: config.normallyClosedRelays,
      normallyClosedLaneSensors: config.normallyClosedLaneSensors,
      useIr: config.useIR,
      debounce: config.debounce,
      numLanes: config.numLanes,
      hasPerLaneRelays: config.hasPerLaneRelays,
      lapPinPitBehavior: config.lapPinPitBehavior,
      lapPinBehaviors: config.lapPinBehaviors,
    };
  }
}
